### PR TITLE
Stereo Imaging Distortion bugfix (SoundScapes Compressor and Limiter link)

### PR DESCRIPTION
**Pull request by JanSto1995**
### What was the problem?
When a compressor or limiter link worked hard (i.e. affected the signal with lots of gain reduction), the stereo sound image got audibly distorted: It seemed like the instruments/the whole mix mix quickly jumped from left to right and vice versa. An audible example was reported by @mikedickey in [this](url) Radio recording.
 
### Assumption of problem origin
The compressor and limiter were compressing stereo channels independently. E.g., if a guitar plays a sound that is panned 80% to the left side, the compressor might only compress channel 1 (the left side of that signal), since channel 2 (the right side of the signal) might not even get louder than the compression threshold. This resulted in the guitar's phantom sound source jumping towards the center of the stereo image, as both channels will be relatively closer to each other in volume.
 
### Implementation of solution (screenshot below)
The [compander.ar() function](url) uses two inputs, one being the actual signal (stereo array) and one being the detection signal (sometimes called side-chain signal).
_(Generally it's cool that we can choose a different signal for loudness detection, as that might allow us to have certain instruments react dynamically to other signals in the long run, allowing for more advanced mixing.)_
Anyways, I've used SuperCollider's [Mix() function](url) to sum our detection signal to mono and have our actual stereo signal duck depending on that mono sum. This means that we take into consideration the loudness of both left&right channels (not just one of them), while we let the compressor and limiter apply its gain reduction equally to both channels in our actual signal.
 
### My test results
I've tested this with stereo signals, and I've tested on both Input and Output chain. Works seemlessly for me now, but I'd appreciate your review.

**Before:**
Screenshot 2022-10-06 at 16 45 43

**After:**
Screenshot 2022-10-06 at 14 01 10

### CODE REVIEW

**Comment by mikedickey** [comment]
thanks @Jan! why `Mix()` versus using max for the side chain? also note that `signal` may either be a 2-channel array (output signal chain) or an array of 2-channel arrays (input signal chain). I just want to make sure that this would do what we expect under both circumstances.

**Comment by JanSto1995** [comment]
Can you tell me a little bit more about the array of 2-ch arrays on the input chain vs. the fixed 2-ch array on the Output chain? I assume that, on the input side, the number of 2-ch arrays is based on the max participants field (plus headroom for additional clients/Jamulus connections)? I assume both operations would work for those arrays of 2-ch arrays

**Comment by JanSto1995** [comment]
Also - wrong Jan :))
Screenshot 2022-10-06 at 22 03 01

**Comment by mikedickey** [comment]
> Can you tell me a little bit more about the array of 2-ch arrays on the input chain vs. the fixed 2-ch array on the Output chain? I assume that, on the input side, the number of 2-ch arrays is based on the max participants field (plus headroom for additional clients/Jamulus connections)? I assume both operations would work for those arrays of 2-ch arrays

Your assumption is correct.

I was just wondering if mix() would end up causing more limiting & compression than is necessary since it's adding the two signals together? For example, say you have roughly the same levels on both left and right and it's near 50%?

I'm not sure if this is enough validation, but running this
```
a = [1, 3];
ArrayMax.ar(a);
```
in sclang console gives me `[ an OutputProxy, an OutputProxy ]`

And running this
```
b = [[2, 4], [7, 8]];
ArrayMax.ar(b);
```
gives me `[ [ an OutputProxy, an OutputProxy ], [ an OutputProxy, an OutputProxy ] ]`

File: links/CompressorLink/CompressorLink.sc [hunk 1]
```diff
@@ -38,7 +38,8 @@ CompressorLink : Link {
 
     ar { | input, id = "" |
         var signal = input;
-        signal = Compander.ar(signal, signal,
+		var monomix = Mix(input); // Mix signal to mono, preventing stereo imaging distortion (same compression for both channels)
```
Comment (mikedickey) [inline links/CompressorLink/CompressorLink.sc:41]: Looks like some weird indenting here.

Also should we use `Mix.kr()` for control rate?

**Review by mikedickey — CHANGES_REQUESTED** [review]

**Comment by jan** [comment]
Glad to see that you have this under control. Keep up the good work!
-- Wrong Jan

**Comment by JanSto1995** [comment]
You're correct about the sum usually being louder than the independent channels, at least if the mid signal (channel 1 & 2 equivalent) is louder than the side signal (channel 1 & 2 phase-inverted [and thus cancelling each other out]). But, generally speaking, the mid signal is louder than the side signal, especially when we assume that most of our sessions are using a pan link which firstly sums to mono. I don't think it would much effect our presets, as I've set compression thresholds rather conservatively with low ratios. But let me double-check that later this week. And it's also worth comparing to your Array.Max() suggestion. Gonna get back to you here.

**Comment by mikedickey** [comment]
@JanSto1995 we noticed today that `[GateLink](https://github.com/jacktrip/jacktrip-sc/blob/main/links/GateLink/GateLink.sc)` is also using `Compander` and likely needs the same modifications.

**Review by mikedickey — APPROVED** [review]
I'm going to approve and merge this since testing results indicate it's a big improvement. The outstanding issues are mainly just additional refinements, so please open another pull request for anything that may surface from additional investigation & testing.

**Comment by JanSto1995** [comment]
@mikedickey sounds good. I'm going to work on your suggested refinements once I have more time. Yes, I'm going to create a new PR then.
I've also just double-checked this push on bleeding edge, and the compressor and limiter are working well with more than one participant.

**Pull request merged**